File: 1301-1400/1319-Number of Operations to Make Network Connected/1319-Number of Operations to Make Network Connected.py
class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if len(connections) < n - 1:
            return -1
        parent = [i for i in range(n)]
        def findParent(x):
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x
        count = n - 1
        for a, b in connections:
            pa = findParent(a)
            pb = findParent(b)
            if pa != pb:
                parent[pb] = pa
                count -= 1
        
        return count

File: 1301-1400/1319-Number of Operations to Make Network Connected/1319-Number of Operations to Make Network Connected.cpp
class Solution {
public:
    int makeConnected(int n, vector<vector<int>>& connections) {
        if (connections.size() < n - 1) {
            return -1;
        }
        int count = n - 1;
        vector<int> parent(n);
        for (int i = 1; i < n; ++i) {
            parent[i] = i;
        }
        for (auto c : connections) {
            int pa = findParent(parent, c[0]);
            int pb = findParent(parent, c[1]);
            if (pa != pb) {
                --count;
                parent[pb] = pa;
            }
        }
        
        return count;
    }
    
private:
    int findParent(vector<int>& parent, int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
};
